resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/rest/kubernetes/PemsToKeyStoreConverter.scala

```diff
@@ -0,0 +1,118 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.deploy.rest.kubernetes
+
+import java.io.{File, FileInputStream, FileOutputStream, InputStreamReader}
+import java.nio.file.Paths
+import java.security.{KeyStore, PrivateKey}
+import java.security.cert.X509Certificate
+import java.util.UUID
+
+import com.google.common.base.Charsets
+import org.bouncycastle.asn1.pkcs.PrivateKeyInfo
+import org.bouncycastle.cert.X509CertificateHolder
+import org.bouncycastle.cert.jcajce.JcaX509CertificateConverter
+import org.bouncycastle.openssl.{PEMKeyPair, PEMParser}
+import org.bouncycastle.openssl.jcajce.JcaPEMKeyConverter
+
+import org.apache.spark.SparkException
+import org.apache.spark.util.Utils
+
+private[spark] object PemsToKeyStoreConverter {
+
+  /**
+   * Loads the given key-cert pair into a temporary keystore file. Returns the File pointing
+   * to where the keyStore was written to disk.
+   */
+  def convertPemsToTempKeyStoreFile(
+      keyPemFile: File,
+      certPemFile: File,
+      keyAlias: String,
+      keyStorePassword: Option[String],
+      keyPassword: Option[String],
+      keyStoreType: Option[String]): File = {
+    require(keyPemFile.isFile, s"Key pem file provided at ${keyPemFile.getAbsolutePath}" +
+      " does not exist or is not a file.")
+    require(certPemFile.isFile, s"Cert pem file provided at ${certPemFile.getAbsolutePath}" +
+      " does not exist or is not a file.")
+    val privateKey = parsePrivateKeyFromPemFile(keyPemFile)
+    val certificate = parseCertificateFromPemFile(certPemFile)
+    val resolvedKeyStoreType = keyStoreType.getOrElse(KeyStore.getDefaultType)
+    val keyStore = KeyStore.getInstance(resolvedKeyStoreType)
+    keyStore.load(null, null)
+    keyStore.setKeyEntry(
+      keyAlias,
+      privateKey,
+      keyPassword.map(_.toCharArray).orNull,
+      Array(certificate))
+    val keyStoreOutputPath = Paths.get(s"keystore-${UUID.randomUUID()}.$resolvedKeyStoreType")
+    Utils.tryWithResource(new FileOutputStream(keyStoreOutputPath.toFile)) { storeStream =>
+      keyStore.store(storeStream, keyStorePassword.map(_.toCharArray).orNull)
+    }
+    keyStoreOutputPath.toFile
+  }
+
+  def convertCertPemToTrustStore(
+      certPemFile: File,
+      certAlias: String,
+      trustStoreType: Option[String]): KeyStore = {
+    require(certPemFile.isFile, s"Cert pem file provided at ${certPemFile.getAbsolutePath}" +
+      " does not exist or is not a file.")
+    val trustStore = KeyStore.getInstance(trustStoreType.getOrElse(KeyStore.getDefaultType))
+    trustStore.load(null, null)
+    val certificate = parseCertificateFromPemFile(certPemFile)
+    trustStore.setCertificateEntry(certAlias, certificate)
+    trustStore
+  }
+
+  private def parsePrivateKeyFromPemFile(keyPemFile: File): PrivateKey = {
+    Utils.tryWithResource(new FileInputStream(keyPemFile)) { keyPemStream =>
```

Comment:
Can you extract it to another function. This triply nested stream seems to be repeating